### PR TITLE
Sensitivity analysis

### PR DESCRIPTION
**Pull request by imgeorgiev**
- Add script `train_wm.py` which focuses on training world models using tdmpc
- add `landscapes.py` which produces figures for analyzing the actor loss landscape for different parameter petributions
- add `smoothness.py` which produces figures for analyzing model smoothness by looking at dynamics jacobian norms and eigenvalues. Note this has a limitation right now where it looks only at first 11 dimensions. Need to fix that but later.

### CODE REVIEW

**Review by varungiridhar — APPROVED** [review]

File: wmlab/config.yaml [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+enc_lr_scale: 0.3
+grad_clip_norm: 20
+actor_grad_clip_norm: 20
+tau: 0.01
+discount_denom: 5
+discount_min: 0.95
+discount_max: 0.995
+buffer_size: 1_000_000 # minimum must be 1000 x num_envs
+exp_name: default
+data_dir: /storage/home/hcoda1/7/igeorgiev3/git/FoWM/data/dflex-ant
```
Comment (varungiridhar) [inline wmlab/config.yaml:42]: filepath on open-source repo is okay?

File: wmlab/landscapes.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    seeding(42, True)
+    env = AntEnv(num_envs=1, no_grad=False, early_termination=False, device=device)
+    actor = ActorStochasticMLP(
+        env.num_obs, env.num_acts, [400, 200, 100], torch.nn.Mish
+    ).to(device)
+    actor2 = ActorStochasticMLP(
+        env.num_obs, env.num_acts, [400, 200, 100], torch.nn.Mish
+    ).to(device)
+    chkpt = torch.load(
+        "/storage/home/hcoda1/7/igeorgiev3/git/FoRL/scripts/outputs/2024-03-14/15-28-46/logs/best_policy.pt",
```
Comment (varungiridhar) [inline wmlab/landscapes.py:39]: filepath, disregard is it's ok

File: wmlab/landscapes.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    from common.parser import parse_cfg
+    from common.math import two_hot_inv
+
+    cfg = parse_cfg(cfg)
+    cfg.obs_shape = {"state": env.obs_space.shape}
+    cfg.action_dim = env.act_space.shape[0]
+    cfg.episode_length = env.episode_length
+    tdmpc = TDMPC2(cfg)
+    tdmpc.load(
+        "/storage/home/hcoda1/7/igeorgiev3/git/FoWM/wmlab/logs/dflex-ant/1/default/models/20000.pt"
```
Comment (varungiridhar) [inline wmlab/landscapes.py:56]: filepath, disregard if okay

File: wmlab/smoothness.py [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+
+    device = "cuda"
+
+    seeding(42, True)
+    env = AntEnv(num_envs=1, no_grad=False, early_termination=False, device=device)
+    actor = ActorStochasticMLP(
+        env.num_obs, env.num_acts, [400, 200, 100], torch.nn.Mish
+    ).to(device)
+    chkpt = torch.load(
+        "/storage/home/hcoda1/7/igeorgiev3/git/FoRL/scripts/outputs/2024-03-14/15-28-46/logs/best_policy.pt",
```
Comment (varungiridhar) [inline wmlab/smoothness.py:44]: filepath, disregard if okay

File: wmlab/smoothness.py [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+    cfg.action_dim = env.act_space.shape[0]
+    cfg.episode_length = env.episode_length
+    tdmpc = TDMPC2(cfg)
+
+    # H: model_path
+    models = {
+        4: "/storage/home/hcoda1/7/igeorgiev3/git/FoWM/wmlab/logs/dflex-ant/1/default/models/20000.pt",
+        8: "/storage/home/hcoda1/7/igeorgiev3/git/FoWM/wmlab/logs/dflex-ant/5/default/models/20000.pt",
+        16: "/storage/home/hcoda1/7/igeorgiev3/git/FoWM/wmlab/logs/dflex-ant/6/default/models/20000.pt",
+        32: "/storage/home/hcoda1/7/igeorgiev3/git/FoWM/wmlab/logs/dflex-ant/7/default/models/20000.pt",
```
Comment (varungiridhar) [inline wmlab/smoothness.py:65]: filepath, disregard if okay

File: wmlab/utils.py [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    from common.parser import parse_cfg
+    from common.math import two_hot_inv
+
+    cfg = parse_cfg(cfg)
+    cfg.obs_shape = {"state": env.obs_space.shape}
+    cfg.action_dim = env.act_space.shape[0]
+    cfg.episode_length = env.episode_length
+    tdmpc = TDMPC2(cfg)
+    tdmpc.load(
+        "/storage/home/hcoda1/7/igeorgiev3/git/FoWM/wmlab/logs/dflex-ant/1/default/models/20000.pt"
```
Comment (varungiridhar) [inline wmlab/utils.py:92]: filepath, disregard if okay

**Review by varungiridhar — APPROVED** [review]

File: wmlab/config.yaml (hunk 1 above)
Comment (imgeorgiev) [reply wmlab/config.yaml:42]: Generally we should leave a placeholder for this in the final public version of this. However, for the time being and for moving fast, this is fine as long as we're on the same machine.
